components/builder-api/src/http/handlers.rs

```diff
@@ -163,18 +172,37 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
                 do_channel_creation(req, ident.get_origin(), &channel, session_id)?;
             }
 
-            do_promotion(req, ident, &channel, session_id)?;
+            let mut project_list = origin_map.entry(ident.get_origin().to_string()).or_insert(
+                Vec::new(),
+            );
+            project_list.push(project);
         } else {
-            projects.push(project.get_ident().to_string());
+            failed_projects.push(project.get_ident().to_string());
+        }
+    }
+
+    // Now that we've sorted all the projects into a HashMap keyed on origin, we can process them,
+    // one origin at a time. We do "core" first.
+    if let Some(core_projects) = origin_map.remove("core") {
+        let promote_result = do_group_promotion(&channel, core_projects, "core");
+        if !promote_result.is_ok() {
+            return promote_result;
+        }
+    }
+
+    for (origin, projects) in origin_map.iter() {
+        let promote_result = do_group_promotion(&channel, projects.to_vec(), &origin);
+        if !promote_result.is_ok() {
+            return promote_result;
```

Comment:
I think it's OK that it feels messy or weird because this is a fundamentally different operation than we've done anywhere else. Here, we're dealing with something that could be a "partial failure".

Thinking a bit more on it, though, I don't know that trying to promote everything we can makes any more sense than stopping at the first failure. If we could access the dependency graph of the packages and try and promote as much as we could in _dependency order_, then that seems like a good approach to take, but without that, maybe the code as-is is how we should go.